backend/runner/observability/runner.go

```diff
@@ -0,0 +1,95 @@
+package observability
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/noop"
+
+	ftlv1 "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1"
+)
+
+const (
+	runnerMeterName              = "ftl.runner"
+	runnerDeploymentKeyAttribute = "ftl.deployment.key"
+	runnerStateNameAttribute     = "ftl.runner.state.name"
+)
+
+type RunnerMetrics struct {
+	meter                  metric.Meter
+	startupFailures        metric.Int64Counter
+	registrationHeartbeats metric.Int64Counter
+	registrationFailures   metric.Int64Counter
+}
+
+func initRunnerMetrics() (*RunnerMetrics, error) {
+	result := &RunnerMetrics{}
+
+	var errs error
+	var err error
+
+	result.meter = otel.Meter(runnerMeterName)
+
+	counter := fmt.Sprintf("%s.startup.failures", runnerMeterName)
+	if result.startupFailures, err = result.meter.Int64Counter(
+		counter,
+		metric.WithDescription("the number of runner startup failures")); err != nil {
+		result.startupFailures, errs = handleInitErrors(counter, err, errs)
+	}
+
+	counter = fmt.Sprintf("%s.registration.heartbeats", runnerMeterName)
+	if result.registrationHeartbeats, err = result.meter.Int64Counter(
+		counter,
+		metric.WithDescription("the number of successful runner (re-)registrations")); err != nil {
+		result.registrationHeartbeats, errs = handleInitErrors(counter, err, errs)
+	}
+
+	counter = fmt.Sprintf("%s.registration.failures", runnerMeterName)
+	if result.registrationFailures, err = result.meter.Int64Counter(
+		counter,
+		metric.WithDescription("the number of failures encounter while attempting to runner registration")); err != nil {
+		result.registrationFailures, errs = handleInitErrors(counter, err, errs)
+	}
+
+	return result, errs
+}
+
+func (m *RunnerMetrics) Registered(ctx context.Context, key *string, state ftlv1.RunnerState) {
+	keyAttr := "unknown"
+	if key != nil {
+		keyAttr = *key
+	}
```

Comment:
We should ideally make `key` an `optional.Option` instead of using a pointer here. Same below. It's annoying because we do have `key` as a pointer already in the runner code, but going forward it's better to avoid pointers to represent nullability :/

Reply:
that looks much better, thanks for the suggestion